network.go

```diff
@@ -504,3 +646,161 @@ func (n *network) getController() *controller {
 	defer n.Unlock()
 	return n.ctrlr
 }
+
+func (n *network) ipamAllocate() ([]func(), error) {
+	var (
+		cnl []func()
+		err error
+	)
+
+	// For now also exclude bridge from using new ipam
+	if n.Type() == "host" || n.Type() == "null" || n.Type() == "bridge" {
+		return cnl, nil
+	}
+
+	ipam, err := n.getController().getIpamDriver(n.ipamType)
+	if err != nil {
+		return nil, err
+	}
+
+	if n.addrSpace == "" {
+		if n.addrSpace, err = n.deriveAddressSpace(); err != nil {
+			return nil, err
+		}
+	}
+
+	if n.ipamV4Config == nil {
+		n.ipamV4Config = []*IpamConf{&IpamConf{}}
+	}
+
+	n.ipamV4Info = make([]*IpamInfo, len(n.ipamV4Config))
+
+	for i, cfg := range n.ipamV4Config {
+		if err = cfg.Validate(); err != nil {
+			return nil, err
+		}
+		d := &IpamInfo{}
+		n.ipamV4Info[i] = d
+
+		d.PoolID, d.Pool, d.Meta, err = ipam.RequestPool(n.addrSpace, cfg.PreferredPool, cfg.SubPool, cfg.Options, cfg.IsV6)
+		if err != nil {
+			return nil, err
+		}
+
+		defer func() {
+			if err != nil {
+				if err := ipam.ReleasePool(d.PoolID); err != nil {
+					log.Warnf("Failed to release address pool %s after failure to create network %s (%s)", d.PoolID, n.Name(), n.ID())
+				}
+			}
+		}()
+
+		if gws, ok := d.Meta[netlabel.Gateway]; ok {
+			if d.Gateway, err = types.ParseCIDR(gws); err != nil {
+				return nil, types.BadRequestErrorf("failed to parse gateway address (%v) returned by ipam driver: %v", gws, err)
+			}
+		}
+
+		// If user requested a specific gateway, libnetwork will allocate it
+		// irrespective of whether ipam driver returned a gateway already.
+		// If none of the above is true, libnetwork will allocate one.
+		if cfg.Gateway != "" || d.Gateway == nil {
```

Comment:
It's ok becasue net.ParseIP will be called on an empty string which results on a  <nil>. Which is the way to tell the IPAM driver to pick any IP.

As the comment says, the gateway IP will always be provided to the network driver. It is the network's gateway. The network driver is free to not use it if it does not need it.